### DISCUSSION TITLE
Increasing Performance for Large Graphs

### DISCUSSION THREAD

**Discussion by bitGatito** [reaviz/reagraph#254] [General]
Visualizing large graphs on Reagraph still leads to performance issues. It noticeably lags when you drag the scene around: [example](https://reagraph.dev/?path=/story/demos-use-cases--mitre-all-techniques). Any idea on how to improve performance?

**Comment by kitswas** [comment]
Experiencing the same here.  
https://citestat.vercel.app/doi/10.48550%2FarXiv.1706.03762  
Corresponding code is at https://github.com/iiith-cs24-sp/citestat/blob/56e9eaa08fdc7b3ab79d137b7112f9ed868b5f42/client/src/components/NetworkGraph.tsx

Reagraph Devs, please consider using web workers as suggested by [Web.dev](https://web.dev/articles/off-main-thread).

As you can see from [this profile](https://share.firefox.dev/3OAlNx7), Reagraph does all of its work on the main (UI) thread, sometimes blocking it for over 2 seconds.

**Reply by amcdnl** [reply]
Happy to accept a PR for this. The edges are super expensive to render so I'm not suprised thats the problem child here.

**Reply by kitswas** [reply]
Another option could be to use [OffscreenCanvas](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas) for rendering.

**Reply by kitswas** [reply]
@amcdnl
I cloned the repo and looked at the code.

Here's the plan:

1. **Improvements:** Functions that iterate over all nodes or edges should utilize Workers.
   Note that this won't decrease the time required to execute those functions, but the time will be consumed on a different thread, leaving the main thread free to do more important tasks like painting (and keep the UI responsive, no 'freeze'). (Total benefit unknown, needs comparision)
2. **API Changes:** Functions in the Reagraph Library that use Workers will be `async` and return `Promise`s. (Possible Breaking Change)
3. **Limitations:** Not all functions can be moved. DOM manipulation, for example, is not available in Workers.
4. [Comlink](https://www.npmjs.com/package/comlink) will be brought in to reduce complexity.

This migration would be best handled by someone very familiar with the codebase.
I am no expert in Web development, so please point out flaws in the plan and possible alternatives.

**Reply by amcdnl** [reply]
The easiest place to start implementing this would be in the layout. The harder place to implement this is going to be in the edge logic which is quite complicated and branches based on number of your edges for performance reasons.
